src/Main.elm

```diff
@@ -0,0 +1,139 @@
+module Main exposing (..)
+
+import Html exposing (..)
+import Html.App as App
+import Html.Attributes exposing (..)
+import Html.Events exposing (..)
+import Json.Decode as Json
+import String
+
+
+-- main
+
+main : Program Never
+main =
+  App.beginnerProgram
+    { model = emptyModel
+    , view = view
+    , update = update
+    }
+
+-- Model
+
+type alias Model =
+  { entries : List Entry
+  , input : String
+  , uid : Int
+  }
+
+type alias Entry =
+  { description : String
+  , id : Int
+  , taps : Int
+  }
+
+emptyModel : Model
+emptyModel =
+  { entries = []
+  , input = ""
+  , uid = 0
+  }
+
+
+-- Update
+
+type Msg
+  = NoOp
+  | UpdateInput String
+  | Add
+  | Tap Int
+
+
+update : Msg -> Model -> Model
+update msg model =
+  case msg of
+    NoOp -> model
+
+    UpdateInput str ->
+      { model | input = str }
+
+    Add ->
+      { model
+        | input = ""
+        , uid = model.uid + 1
+        , entries =
+            if String.isEmpty model.input then
+              model.entries
+            else
+              model.entries ++ [newEntry model.input model.uid]}
+
+    Tap id' ->
+      let updateEntry e =
+        if e.id == id' then { e | taps = e.taps + 1 } else e
+      in
+        { model | entries = List.map updateEntry model.entries }
+
+
+newEntry : String -> Int -> Entry
+newEntry desc id' =
+  { description = desc
+  , id = id'
+  , taps = 0
+  }
+
+onEnter : Msg -> Attribute Msg
+onEnter msg =
+  let
+    tagger code =
+      if code == 13 then msg else NoOp
+  in
+    on "keydown" (Json.map tagger keyCode)
```

Comment:
OK, honestly, I just cribbed this straight from the TodoMVC implementation (cut me some slack: it was hot out and I was feeling sick), so I had no idea what was going on when I wrote this. :)

Here are the type signatures we're looking at:

    on : String -> Json.Decode.Decoder a -> Html.Attribute a
    keyCode : Json.Decode.Decoder Int
    Json.map : (a -> b) -> Json.Decode.Decoder a -> Json.Decode.Decoder b

To grok `map`, it may or may not be helpful to read about [Haskell functors](http://learnyouahaskell.com/functors-applicative-functors-and-monoids), since it looks like `Decoder a` is a functor.

Here's [the docs for `on`](http://package.elm-lang.org/packages/evancz/elm-html/4.0.1/Html-Events#on), which is where it looks like the magic is happening. Looks like you give it: 1) an event name, 2) a decoder that gives you access to the event's data, then 3) a function that does something with the event data; and it gives you back an `Html.Attribute` that you can plug into an HTML-generating function's attribute list.

So I guess something in the `Html` library is doing the work of binding the event at the appropriate place in the tree—the place, I guess, where you plug the output of `on` into an HTML-generating function.